content/posts/messaging-infrastructure.md

```diff
@@ -0,0 +1,80 @@
+# Messaging infrastructure
+
+So far we've seen message de-duplication is not a simple topic and that *idempotency* is not a useful term then talking about distributed systems in which messages can get duplicated and reordered. Since we are sending our readers down the path of complexity we should at least discuss briefly what are the alternatives.
+
+## Building blocks
+
+In the model of a distributed system we use, all business processes are implemented by combining one simple building block. That building block consists of five operations
+- receive message
+- execute business logic
+- persist state change
+- send outgoing messages
+- consume the receive message 
+Note that in the previous posts we've shown how the these blocks have slightly different shape at the system boundary. The behavior of the system depends greatly on how these five operations comprising a block are combined into atomic transactions. 
+
+## Retries
+
+Throughout this article we are assuming that the message queue we are using is durable and it re-delivers messages for processing in case previous processing attempts have failed. Receiving of a message is a read operation that has no side effects. In order to remove a message from the queue it needs to be consumed.
+
+## Scope of a transaction
+
+If two operations are combined into a transactions, it is guaranteed that either both operations complete successfully or neither is. This is the atomicity property of a transaction. Atomicity is not enough, though. We also need to make sure that the results of the transaction are not lost. In other words, we need these transactions to also be durable. That means that an operation can be transactional if it interacts with a medium that can durably record state. 
+
+Let's see how the five operations of our basic building block can be combined to form transactions.
+
+## No transactions
```

Comment:
Should we say that for each we want to achieve consistent messaging as defined in previous posts?